skrifa/src/color/mod.rs

```diff
@@ -0,0 +1,127 @@
+//! Color outline support.
+
+// Remove me when code is filled in
+#![allow(unused_variables, dead_code)]
+
+mod instance;
+
+use core::ops::Range;
+
+use crate::{instance::LocationRef, metrics::BoundingBox};
+use read_fonts::{tables::colr, types::GlyphId, ReadError, TableProvider};
+
+use instance::{resolve_paint, PaintId};
+
+/// Interface for receiving a sequence of commands that represent a flattened
+/// paint graph in a color outline.
+// Placeholder
+pub trait ColorPainter {}
+
+/// Affine transformation matrix.
+#[derive(Copy, Clone, Debug)]
+pub struct Transform {
+    xx: f32,
+    yx: f32,
+    xy: f32,
+    yy: f32,
+    dx: f32,
+    dy: f32,
+}
+
+/// Reference to a paint graph that represents a color glyph outline.
+#[derive(Clone)]
+pub struct ColorOutline<'a> {
+    colr: colr::Colr<'a>,
+    glyph_id: GlyphId,
+    kind: ColorOutlineKind<'a>,
+}
+
+impl<'a> ColorOutline<'a> {
+    /// Returns the glyph identifier that was used to retrieve this outline.
+    pub fn glyph_id(&self) -> GlyphId {
+        self.glyph_id
+    }
+
+    /// Evaluates the paint graph at the specified location in variation space
+    /// and returns the bounding box for the full scene.
+    ///
+    /// The `glyph_bounds` closure will be invoked for each clip node in the
+    /// graph and should return the bounding box for the given glyph, location
+    /// and transform.
+    pub fn bounding_box(
+        &self,
+        location: impl Into<LocationRef<'a>>,
+        glyph_bounds: impl FnMut(GlyphId, &LocationRef, Transform) -> Option<BoundingBox>,
```

Comment:
In trying to understand your intention, first some questions: So the difference from a "degenerate" `ColorPainter` implementation on the client side (to build a bounding box there) would be that the transform matrix is accumulated internally and we would call back less frequently, not for each node in the graph? Since there is no mention in the comment about what to do with compositions where the path through the graph forks, do you mean this to track the accumulated transform internally, and call back for each clip node with a fully spelled-out transform? 

A previously configured, client-side transform (on the viewport for example) would be taken into account by the client multiplying that in when it receives the callback? 

So the client would perform own path retrieval, combine its own possible viewport transform with the accumulated transform from the callback, and compute the bounding box of the path, with transform applied, and return that? 

I do appreciate that this would disconnect the color module from the scale module - which I think make sense, as we discussed in the review of 

This works as an interface, and we can shift slightly more code to the Rust side. I am trying to weigh whether this somewhat unusual API pattern is worth defining in addition to the to the paint traversal call, with which the client can compute a bounding box as well: it receives more callback and the matrix tracking would need to happen on the client side, but since we have that on the 2D graphics library side - that's doable there just as well.

Reply:
The answer to all of your questions is “yes” with the exception that this wouldn’t do any caching of paths internally although the client could certainly do so if they’d like to trade space for time. Seems to be a point in favor of this approach?

I hadn’t considered a client provided root transform. We can accept that as a parameter here and fold it in if you think that’d be more useful. Otherwise, yes, a client could apply it in the callback.

I have two general concerns:

1. Handling edge cases (composition nodes, detecting unbounded subgraphs) is not exactly trivial so it would be nice to offer a correct implementation of this instead of pushing that burden to all potential clients. 

2. While not a top priority atm, I’m trying to keep the embedded case in mind wrt heap allocations. This interface would allow the caller to decide how to handle memory requirements for loading glyphs. Aside: that does leave the recursion blacklist as a potential problem— Rod and I discussed adding a cfg feature that uses a max recursion depth instead but that’s not something we need to address now.